libraries/libraries.go

```diff
@@ -0,0 +1,208 @@
+}
+
+// Get implements the Library interface.
+func (l *Libraries) Get(id borges.RepositoryID, mode borges.Mode) (borges.Repository, error) {
+	var repos []borges.Repository
+	for _, lib := range l.libs {
+		r, err := lib.Get(id, mode)
+		if err != nil {
+			if borges.ErrRepositoryNotExists.Is(err) {
+				continue
+			}
+
+			return nil, err
+		}
+
+		repos = append(repos, r)
+	}
+
+	if len(repos) == 0 {
+		return nil, borges.ErrRepositoryNotExists.New(id)
+	}
+
+	return repos[0], nil
+}
+
+// GetOrInit implements the Library interface.
+func (l *Libraries) GetOrInit(borges.RepositoryID) (borges.Repository, error) {
+	return nil, borges.ErrNotImplemented.New()
+}
+
+// Has implements the Library interface.
+func (l *Libraries) Has(id borges.RepositoryID) (bool, borges.LibraryID, borges.LocationID, error) {
+	var matches []*hasReturnValues
+	for _, lib := range l.libs {
+		has, libID, locID, err := lib.Has(id)
+		if err != nil {
+			return false, "", "", err
+		}
+
+		if has {
+			matches = append(matches, &hasReturnValues{
+				libID: libID,
+				locID: locID,
+			})
+		}
+	}
+
+	if len(matches) == 0 {
+		return false, "", "", nil
+	}
+
+	return true, matches[0].libID, matches[0].locID, nil
+}
+
+type hasReturnValues struct {
+	libID borges.LibraryID
+	locID borges.LocationID
+}
+
+// Repositories implements the Library interface.
+func (l *Libraries) Repositories(mode borges.Mode) (borges.RepositoryIterator, error) {
+	var repositories []borges.RepositoryIterator
+	for _, lib := range l.libs {
+		repos, err := lib.Repositories(mode)
+		if err != nil {
+			return nil, err
+		}
+
+		repositories = append(repositories, repos)
+	}
+
+	return MergeRepositoryIterators(repositories), nil
+}
+
+// Location implements the Library interface.
+func (l *Libraries) Location(id borges.LocationID) (borges.Location, error) {
+	var locations []borges.Location
+	for _, lib := range l.libs {
+		loc, err := lib.Location(id)
+		if err != nil {
+			if borges.ErrLocationNotExists.Is(err) {
+				continue
+			}
+
+			return nil, err
+		}
+
+		locations = append(locations, loc)
+	}
+
+	if len(locations) == 0 {
+		return nil, borges.ErrLocationNotExists.New(id)
+	}
+
+	return locations[0], nil
+}
+
+// Locations implements the Library interface.
+func (l *Libraries) Locations() (borges.LocationIterator, error) {
+	var locations []borges.LocationIterator
+	for _, lib := range l.libs {
+		locs, err := lib.Locations()
+		if err != nil {
+			return nil, err
+		}
+
+		locations = append(locations, locs)
+	}
+
+	return MergeLocationIterators(locations), nil
+}
+
+// Library implements the Library interface.
+func (l *Libraries) Library(id borges.LibraryID) (borges.Library, error) {
+	lib, ok := l.libs[id]
+	if !ok {
+		return nil, borges.ErrLibraryNotExists.New(id)
+	}
+
+	return lib, nil
+}
+
+// Libraries implements the Library interface.
+func (l *Libraries) Libraries() (borges.LibraryIterator, error) {
+	return l.FilteredLibraries(func(borges.Library) (bool, error) {
```

Comment:
What is the use of `FilteredLibraries`?

Reply:
It allows you to filter libraries, for example by their types:
```go
	var filter FilterLibraryFunc = func(lib borges.Library) (bool, error) {
		_, ok := lib.(*plain.Library)
		return ok, nil
	}

	iter, err := s.libs.FilteredLibraries(filter)
```
This mechanism along with the `Merge(Repository/Location/Library)Iterator` functions make possible build and iterator with whatever scheduling plan you want.

You could filter by `siva` library type -> get a repository iterator from the first location from the first library, get a repository iterator from the first location from the second library, etc. Merge them with `MergeRepositoryIterator` so the order is kept.

Do another  scheduling plan for `plain` libraries ...

Merge again the iterators.

If we need very specifics predefined plan we can add a couple of functions to return those specific iterators using this mechanism.

Reply:
Can you add an example for the plan defined in the issue? Also, does this mean that we have to create one iterator per location before starting to iterate?

Reply:
Example:
```go
func coolSivaRepositoryIterator(libs *Libraries, mode borges.Mode) (borges.RepositoryIterator, error) {
	var filter FilterLibraryFunc = func(lib borges.Library) (bool, error) {
		_, ok := lib.(*siva.Library)
		return ok, nil
	}

	libIter, err := libs.FilteredLibraries(filter)
	if err != nil {
		return nil, err
	}

	var locsIter []*closedLocIter
	err = libIter.ForEach(func(lib borges.Library) error {
		locIter, err := lib.Locations()
		if err == nil {
			locsIter = append(locsIter, &closedLocIter{locIter, false})
		}

		return err
	})

	if err != nil {
		return nil, err
	}

	var repos []borges.RepositoryIterator
	for !areClosed(locsIter) {
		for _, li := range locsIter {
			loc, err := li.Next()
			if err != nil {
				if err == io.EOF {
					li.Close()
					continue
				}

				return nil, err
			}

			ri, err := loc.Repositories(mode)
			if err != nil {
				return nil, err
			}

			repos = append(repos, ri)
		}
	}

	return MergeRepositoryIterators(repos), nil
}

func areClosed(locs []*closedLocIter) bool {
	if len(locs) == 0 {
		return true
	}

	for _, loc := range locs {
		if !loc.closed {
			return false
		}
	}

	return true
}

type closedLocIter struct {
	borges.LocationIterator

	closed bool
}

func (i *closedLocIter) Close() {
	if i.closed {
		return
	}

	i.closed = true
	i.LocationIterator.Close()
}
```

Order:
```
REGULAR ORDER
-------
github.com/diptadhi/datasharing
github.com/nmorr041/datasharing
github.com/jtleek/datasharing
github.com/enaqx/awesome-pentest
github.com/Inter1292/awesome-pentest
github.com/apelsin83/awesome-pentest
github.com/rtyley/small-test-repo
github.com/kuldeep992/small-test-repo
github.com/kuldeep-singh-blueoptima/small-test-repo
github.com/kahun/awesome-sysadmin
github.com/apoliukh/awesome-sysadmin
github.com/gauravaristocrat/awesome-sysadmin
github.com/MunGell/awesome-for-beginners
github.com/dhruvil1514/awesome-for-beginners
github.com/karellism/awesome-for-beginners
github.com/prakhar1989/awesome-courses
github.com/Leo-xxx/awesome-courses
github.com/manjunath00/awesome-courses
github.com/jtoy/awesome-tensorflow
github.com/SiweiLuo/awesome-tensorflow
github.com/youtang1993/awesome-tensorflow

COOL ORDER
----
github.com/jtleek/datasharing
github.com/diptadhi/datasharing
github.com/nmorr041/datasharing
github.com/rtyley/small-test-repo
github.com/kuldeep992/small-test-repo
github.com/kuldeep-singh-blueoptima/small-test-repo
github.com/prakhar1989/awesome-courses
github.com/Leo-xxx/awesome-courses
github.com/manjunath00/awesome-courses
github.com/enaqx/awesome-pentest
github.com/Inter1292/awesome-pentest
github.com/apelsin83/awesome-pentest
github.com/kahun/awesome-sysadmin
github.com/apoliukh/awesome-sysadmin
github.com/gauravaristocrat/awesome-sysadmin
github.com/jtoy/awesome-tensorflow
github.com/SiweiLuo/awesome-tensorflow
github.com/youtang1993/awesome-tensorflow
github.com/MunGell/awesome-for-beginners
github.com/dhruvil1514/awesome-for-beginners
github.com/karellism/awesome-for-beginners
```